File: Go/EonaCatCipher.go
package main

import (
    "crypto/rand"
    "crypto/sha512"
    "crypto/hmac"
    "encoding/binary"
    "encoding/hex"
    "encoding/base64"
    "fmt"
    "log"
)

/*
 * EonaCatCipher - Because security is key!
 * 
 * Copyright (c) 2024 EonaCat (Jeroen Saey)
 * 
 * https://eonacat.com/license
 * 
 *   TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION 
 *                  OF SOFTWARE BY EONACAT (JEROEN SAEY)
 *
 * This software is provided "as is", without any express or implied warranty.
 * In no event shall the authors or copyright holders be liable for any claim,
 * damages or other liability, whether in an action of contract, tort or otherwise,
 * arising from, out of or in connection with the software or the use or other
 * dealings in the software.
 * 
 * You may use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
 * copies of the Software, and permit persons to whom the Software is furnished 
 * to do so, subject to the following conditions:
 * 
 * 1. The above copyright notice and this permission notice shall be included in 
 *    all copies or substantial portions of the Software.
 * 
 * 2. The software must not be used for any unlawful purpose.
 * 
 * For any inquiries, please contact: eonacat@gmail.com
 */

// EonaCatCipher struct
type EonaCatCipher struct {
    derivedKey []byte
    hmacKey    []byte
    ivSize     int
    keySize    int
    rounds     int
    blockSize  int
}

// Constants
const (
    DEFAULT_SALT_SIZE  = 2048 // Salt size for key derivation
    DEFAULT_IV_SIZE    = 2048    // IV size (16384 bits)
    DEFAULT_KEY_SIZE   = 2048    // Key size (16384 bits)
    DEFAULT_ROUNDS     = 2048
    DEFAULT_BLOCK_SIZE = 8192  // 8KB
    HMAC_KEY_SIZE      = 32    // Key size for HMAC (256 bits)
    SECRET_SAUCE       = 0x5DEECE66D
)

// NewEonaCatCipher constructor
func NewEonaCatCipher(password string, saltSize int, ivSize int, keySize int, rounds int, blockSize int) *EonaCatCipher {
    if password == "" {
        log.Fatal("EonaCatCipher: Password cannot be null or empty.")
    }

    ec := &EonaCatCipher{
        ivSize:    ivSize,
        keySize:   keySize,
        rounds:    rounds,
        blockSize: blockSize,
    }

    ec.derivedKey, ec.hmacKey = ec.DeriveKeyAndHMAC(password, saltSize)

    return ec
}

// GenerateRandomBytes generates a slice of random bytes
func GenerateRandomBytes(size int) ([]byte, error) {
    randomBytes := make([]byte, size)
    _, err := rand.Read(randomBytes)
    if err != nil {
        return nil, err
    }
    return randomBytes, nil
}

// DeriveKeyAndHMAC derives the encryption key and HMAC key from the password
func (ec *EonaCatCipher) DeriveKeyAndHMAC(password string, saltSize int) ([]byte, []byte) {
    salt, _ := GenerateRandomBytes(saltSize)
    encryptionKey := ec.PBKDF2(password, salt, ec.keySize, ec.rounds)

    // Derive separate key for HMAC
    hmacKey := ec.PBKDF2(password, salt, HMAC_KEY_SIZE, ec.rounds)

    keyWithSalt := append(salt, encryptionKey...)

    return keyWithSalt, hmacKey
}

// PBKDF2 implementation of the PBKDF2 key derivation function
func (ec *EonaCatCipher) PBKDF2(password string, salt []byte, keyLength int, iterations int) []byte {
    hmacHash := hmac.New(sha512.New, []byte(password))
    hashLength := hmacHash.Size()
    requiredBytes := keyLength
    blocksNeeded := (requiredBytes + hashLength - 1) / hashLength

    derivedKey := make([]byte, requiredBytes)

    for blockIndex := 1; blockIndex <= blocksNeeded; blockIndex++ {
        currentBlock := append(salt, intToBytes(blockIndex)...)

        // U1 = HMAC(password, salt + blockIndex)
        u := hmacHash.Sum(currentBlock)
        block := make([]byte, hashLength)
        copy(block, u)

        // Derived key starts with U1
        copy(derivedKey[(blockIndex-1)*hashLength:], block)

        // Iterations
        for iteration := 1; iteration < iterations; iteration++ {
            u = hmacHash.Sum(u)

            // XOR with previous result
            for i := 0; i < hashLength; i++ {
                block[i] ^= u[i]
            }

            // Append result to derived key
            copy(derivedKey[(blockIndex-1)*hashLength:], block)
        }
    }

    return derivedKey
}

// Encrypt encrypts the plaintext using the derived key and returns the ciphertext with HMAC
func (ec *EonaCatCipher) Encrypt(plaintext string) ([]byte, error) {
    iv, err := GenerateRandomBytes(ec.ivSize)
    if err != nil {
        return nil, err
    }
    plaintextBytes := []byte(plaintext)
    ciphertext := make([]byte, len(plaintextBytes))

    cipher := NewEonaCatCrypto(ec.derivedKey, iv, ec.blockSize, ec.rounds)
    cipher.Generate(plaintextBytes, ciphertext, true)

    // Combine IV and ciphertext
    result := append(iv, ciphertext...)

    // Generate HMAC for integrity check
    hmac := ec.GenerateHMAC(result)

    // Combine result and HMAC
    finalResult := append(result, hmac...)

    return finalResult, nil
}

// Decrypt decrypts the ciphertext with HMAC
func (ec *EonaCatCipher) Decrypt(ciphertextWithHMAC []byte) (string, error) {
    hmacOffset := len(ciphertextWithHMAC) - HMAC_KEY_SIZE

    // Separate HMAC from the ciphertext
    providedHMAC := ciphertextWithHMAC[hmacOffset:]
    ciphertext := ciphertextWithHMAC[:hmacOffset]

    // Verify HMAC before decrypting
    calculatedHMAC := ec.GenerateHMAC(ciphertext)
    if !AreEqual(providedHMAC, calculatedHMAC) {
        return "", fmt.Errorf("EonaCatCipher: HMAC validation failed. Data may have been tampered with.")
    }

    // Extract IV
    iv := ciphertext[:ec.ivSize]

    // Extract encrypted data
    encryptedData := ciphertext[ec.ivSize:]

    // Decrypt
    decryptedData := make([]byte, len(encryptedData))
    cipher := NewEonaCatCrypto(ec.derivedKey, iv, ec.blockSize, ec.rounds)
    cipher.Generate(encryptedData, decryptedData, false)

    return string(decryptedData), nil
}

// GenerateHMAC generates HMAC for the data
func (ec *EonaCatCipher) GenerateHMAC(data []byte) []byte {
    h := hmac.New(sha512.New, ec.hmacKey)
    h.Write(data)
    return h.Sum(nil)
}

// AreEqual checks if two byte slices are equal
func AreEqual(a, b []byte) bool {
    if len(a) != len(b) {
        return false
    }
    for i := range a {
        if a[i] != b[i] {
            return false
        }
    }
    return true
}

// Convert an integer to bytes
func intToBytes(n int) []byte {
    buf := make([]byte, 4)
    binary.BigEndian.PutUint32(buf, uint32(n))
    return buf
}

// EonaCatCrypto struct for encryption and decryption
type EonaCatCrypto struct {
    blockSize    int
    rounds       int
    state        []uint64
    key          []uint32
    nonce        []uint32
    blockCounter uint32
}

// NewEonaCatCrypto constructor
func NewEonaCatCrypto(keyWithSalt []byte, nonce []byte, blockSize int, rounds int) *EonaCatCrypto {
    ec := &EonaCatCrypto{
        blockSize: blockSize / 4,
        rounds:    rounds,
        key:       make([]uint32, len(keyWithSalt)/4),
        nonce:     make([]uint32, len(nonce)/4),
        state:     make([]uint64, blockSize/4),
    }
    copy(ec.key, keyWithSalt)
    copy(ec.nonce, nonce)
    return ec
}

// GenerateBlock generates a block for encryption/decryption
func (ec *EonaCatCrypto) GenerateBlock(output []byte) {
    // Initialize state
    for i := 0; i < len(ec.state); i++ {
        ec.state[i] = uint64(ec.key[i%len(ec.key)]) ^ uint64(ec.nonce[i%len(ec.nonce)]) + uint64(i)*SECRET_SAUCE
    }

    // Mix the states according to the rounds
    for round := 0; round < ec.rounds; round++ {
        for i := 0; i < len(ec.state); i++ {
            ec.state[i] = (ec.state[i] + uint64(round)) ^ (uint64(round) * SECRET_SAUCE) + uint64(i) + uint64(ec.blockCounter)
        }
    }

    // Output block
    for i := 0; i < len(output); i++ {
        output[i] = byte(ec.state[i/4] >> (8 * (i % 4)))
    }
    ec.blockCounter++
}

// Generate performs encryption or decryption
func (ec *EonaCatCrypto) Generate(input []byte, output []byte, encrypt bool) {
    totalBlocks := (len(input) + ec.blockSize - 1) / ec.blockSize

    for blockIndex := 0; blockIndex < totalBlocks; blockIndex++ {
        inputOffset := blockIndex * ec.blockSize
        outputOffset := blockIndex * ec.blockSize
        block := make([]byte, ec.blockSize)

        // Generate a block based on the input
        ec.GenerateBlock(block)

        // Perform XOR for encryption or decryption
        for i := 0; i < len(block) && inputOffset+i < len(input); i++ {
            output[outputOffset+i] = input[inputOffset+i] ^ block[i]
        }
    }
}

// Main function for testing
func main() {
    password := "securePassword123!@#$"
    plaintext := "Thank you for using EonaCatCipher!"

    fmt.Printf("Encrypting '%s' with password '%s' (we do this 5 times)\n", plaintext, password)
    fmt.Println("================")

    for i := 0; i < 5; i++ {
        fmt.Printf("Encryption round %d:\n", i+1)
        fmt.Println("================")

        cipher := NewEonaCatCipher(password, DEFAULT_SALT_SIZE, DEFAULT_IV_SIZE, DEFAULT_KEY_SIZE, DEFAULT_ROUNDS, DEFAULT_BLOCK_SIZE)
        encrypted, err := cipher.Encrypt(plaintext)
        if err != nil {
            log.Fatal(err)
        }

        fmt.Printf("Encrypted (base64): %s\n", base64.StdEncoding.EncodeToString(encrypted))

        decrypted, err := cipher.Decrypt(encrypted)
        if err != nil {
            log.Fatal(err)
        }

        fmt.Printf("Decrypted: %s\n", decrypted)
        fmt.Println("================")
    }
}

File: Python/EonaCatCipher.py
import os
import hashlib
import hmac
import struct
import base64

class EonaCatCipher:
    DEFAULT_SALT_SIZE = 2048;     # Salt size for key derivation
    DEFAULT_IV_SIZE = 2048;       # IV size (16384 bits)
    DEFAULT_KEY_SIZE = 2048;      # Key size (16384 bits)
    DEFAULT_ROUNDS = 2048;        # Rounds
    DEFAULT_BLOCK_SIZE = 8192;    # 8kb
    HMAC_KEY_SIZE = 32;           # Key size for HMAC (256 bits)

#/*
# * EonaCatCipher - Because security is key!
# * 
# * Copyright (c) 2024 EonaCat (Jeroen Saey)
# * 
# * https://eonacat.com/license
# * 
# *   TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION 
# *                  OF SOFTWARE BY EONACAT (JEROEN SAEY)
# *
# * This software is provided "as is", without any express or implied warranty.
# * In no event shall the authors or copyright holders be liable for any claim,
# * damages or other liability, whether in an action of contract, tort or otherwise,
# * arising from, out of or in connection with the software or the use or other
# * dealings in the software.
# * 
# * You may use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
# * copies of the Software, and permit persons to whom the Software is furnished 
# * to do so, subject to the following conditions:
# * 
# * 1. The above copyright notice and this permission notice shall be included in 
# *    all copies or substantial portions of the Software.
# * 
# * 2. The software must not be used for any unlawful purpose.
# * 
# * For any inquiries, please contact: eonacat@gmail.com
# */

    def __init__(self, password, salt_size=None, iv_size=None, key_size=None, rounds=None, block_size=None):
        if not password:
            raise ValueError("EonaCatCipher: Password cannot be null or empty.")

        self.salt_size = salt_size if salt_size is not None else self.DEFAULT_SALT_SIZE
        self.iv_size = iv_size if iv_size is not None else self.DEFAULT_IV_SIZE // 8  # Convert bits to bytes
        self.key_size = key_size if key_size is not None else self.DEFAULT_KEY_SIZE // 8  # Convert bits to bytes
        self.rounds = rounds if rounds is not None else self.DEFAULT_ROUNDS
        self.block_size = block_size if block_size is not None else self.DEFAULT_BLOCK_SIZE // 8  # Convert bits to bytes

        # Derive encryption key and HMAC key
        self.derived_key, self.hmac_key = self.derive_key_and_hmac(password)

    @staticmethod
    def generate_random_bytes(size):
        return os.urandom(size)

    def derive_key_and_hmac(self, password):
        salt = self.generate_random_bytes(self.salt_size)
        encryption_key = self.pbkdf2(password, salt, self.key_size, self.rounds)

        # Derive separate key for HMAC
        hmac_key = self.pbkdf2(password, salt, self.HMAC_KEY_SIZE, self.rounds)

        key_with_salt = salt + encryption_key

        return key_with_salt, hmac_key

    @staticmethod
    def pbkdf2(password, salt, key_length, iterations):
        # PBKDF2 using HMAC-SHA512
        hmac_sha512 = hashlib.pbkdf2_hmac('sha512', password.encode(), salt, iterations, dklen=key_length)
        return hmac_sha512

    def encrypt(self, plaintext):
        iv = self.generate_random_bytes(self.iv_size)
        plaintext_bytes = plaintext.encode()

        ciphertext = bytearray(len(plaintext_bytes))

        # Generate cipher
        cipher = EonaCatCrypto(self.derived_key, iv, self.block_size, self.rounds)
        cipher.generate(plaintext_bytes, ciphertext, True)

        # Combine IV and ciphertext
        result = iv + ciphertext

        # Generate HMAC for integrity check
        hmac = self.generate_hmac(result)

        # Combine result and HMAC
        final_result = result + hmac
        return final_result

    def decrypt(self, ciphertext_with_hmac):
        hmac_offset = len(ciphertext_with_hmac) - self.HMAC_KEY_SIZE

        # Separate HMAC from the ciphertext
        provided_hmac = ciphertext_with_hmac[hmac_offset:]
        ciphertext = ciphertext_with_hmac[:hmac_offset]

        # Verify HMAC before decrypting
        calculated_hmac = self.generate_hmac(ciphertext)
        if not self.are_equal(provided_hmac, calculated_hmac):
            raise ValueError("EonaCatCipher: HMAC validation failed. Data may have been tampered with.")

        # Extract IV
        iv = ciphertext[:self.iv_size]
        encrypted_data = ciphertext[self.iv_size:]

        # Decrypt
        decrypted_data = bytearray(len(encrypted_data))
        cipher = EonaCatCrypto(self.derived_key, iv, self.block_size, self.rounds)
        cipher.generate(encrypted_data, decrypted_data, False)

        return decrypted_data.decode()

    def generate_hmac(self, data):
        return hmac.new(self.hmac_key, data, hashlib.sha256).digest()

    @staticmethod
    def are_equal(a, b):
        return hmac.compare_digest(a, b)

class EonaCatCrypto:
    SECRET_SAUCE = 0x5DEECE66D

    def __init__(self, key_with_salt, nonce, block_size, rounds):
        self.rounds = rounds
        self.block_size = block_size // 4 > 0 and block_size // 4 or 128

        self.key = list(struct.unpack(f'>{len(key_with_salt) // 4}I', key_with_salt))
        self.nonce = list(struct.unpack(f'>{len(nonce) // 4}I', nonce))
        self.state = [0] * (self.block_size // 4)
        self.block_counter = 0

    def generate_block(self, output):
        # Initialize state using a combined operation
        for i in range(len(self.state)):
            self.state[i] = (self.key[i % len(self.key)] ^ self.nonce[i % len(self.nonce)]) + (i * self.SECRET_SAUCE)

        # Mix the states according to the rounds
        for round in range(self.rounds):
            for i in range(len(self.state)):
                self.state[i] = (self.state[i] + round) ^ (round * self.SECRET_SAUCE) + (i + self.block_counter)

        # Output block
        output.extend(self.state)
        self.block_counter += 1

    def generate(self, input_data, output, encrypt):
        total_blocks = (len(input_data) + self.block_size - 1) // self.block_size

        for block_index in range(total_blocks):
            input_offset = block_index * self.block_size
            block = bytearray(self.block_size)

            # Generate a block based on the input
            self.generate_block(block)

            # Perform XOR for encryption or decryption
            for i in range(len(block)):
                if input_offset + i < len(input_data):
                    output[input_offset + i] = input_data[input_offset + i] ^ block[i]

def main():
    password = "securePassword123!@#$"
    plaintext = "Thank you for using EonaCatCipher!"

    print(f"Encrypting '{plaintext}' with password '{password}' (we do this 5 times)")
    print("================")

    for i in range(5):
        print(f"Encryption round {i + 1}: ")
        print("================")

        cipher = EonaCatCipher(password)
        encrypted = cipher.encrypt(plaintext)

        print("Encrypted (byte array):", [b for b in encrypted])
        
        decrypted = cipher.decrypt(encrypted)
        print("Decrypted:", decrypted)
        print("================")

if __name__ == "__main__":
    main()
